### ISSUE TITLE
Provider Contexts

### ISSUE THREAD

**Issue by tmehlinger** [pulumi/pulumi#6835]
I've done my best to make up a name that describes the feature I'll propose below. Hopefully it will accurately convey what I'm looking for. 😄 

We have gone to great lengths to configure _everything_ we manage with Pulumi, from basic infrastructure all the way up to applications themselves (in Kubernetes). Pulumi has been great for us because we have a full vertical integration of infra and application config, with the full dependency chain encoded with a single tool.

However, we've run into a few gotchas linking up dependencies when we need to be explicit with providers. One example is delegating DNS records with Route 53. Suppose we have two AWS accounts, A and B. In account A, we bring up a stack (say, `root`) that creates a zone in Route 53 and IAM policy that permits account B to assume a role with permissions to do CRUD on records to delegate a subdomain to itself. In account B, we create a stack (call it `delegate`) to manage the zone to host the subdomain.

The configuration for `delegate` might look something like this:
```typescript
const ref = new pulumi.StackReference('root');
const zoneId = ref.requireOutput('zoneId');

new aws.route53.Record('delegatedZone', {
    zoneId: zoneId,
    ...
}, {
    provider: new aws.Provider(...),
})
```

Passing the provider explicitly is fine when there's only one resource that needs to use an alternate provider but it would be nice if there were a way to make the provider implicit in the scope of some context. I propose something like this:

```typescript
const outputs = aws.withProvider(awsProviderArgs, () => {
    const record = new aws.route53.Record(...);
    return {
        record.id,
    };
});
```

That's obviously a lot of extra machinery for swapping providers on a single resource. However, where it would be _really_ useful is in Kubernetes. We provision EKS clusters using Pulumi and we would like to create application stacks with explicit dependencies on those clusters without making _any_ assumptions about the environment. That is to say, applying configuration for an application should assume _nothing_ about what the environment can provide in order to authenticate with an EKS cluster and bring up resources. Example:
```typescript
# EKS stack
const cluster = new aws.eks.Cluster(...);
export const kubeconfig = createKubeConfig(cluster);


# application stack
const ref = new pulumi.StackReference('eksStack');
const kubeconfig = ref.requireOutput('kubeConfig');

const outputs = k8s.withProvider({ kubeconfig: kubeconfig }, () => {
    new k8s.apps.v1.Deployment(...);
    new k8s.core.v1.Service(...);
    new k8s.networking.v1beta1.Ingress(...);
    return {
        someOutput: 'anOutput',
    };
});
```

In this example, we've eliminated the boilerplate (and potential for bugs) of passing an opts object to every resource.

I'm happy to contrive more example if that would be helpful.

Thank you!

**Comment by leezen** [comment]
@tmehlinger Thanks for the suggestion! While it's not the same syntax, you can somewhat accomplish this today via component resources. Child resources of the parent resource will inherit whatever provider configuration that's provided to the parent. You'll still need to ensure that you parent each child resource when constructing them, but at least you're not trying to juggle potentially multiple `opts` objects.

**Comment by tmehlinger** [comment]
Yep, we had been doing it that way and recently started _undoing_ that because of the occasional mayhem that comes with forgetting to explicitly give the parent to the child resource. We resolved to do things a bit differently in more verbose and repetitive fashion, then return to clean things up in the future if we can get something like this. :)

**Comment by lukehoban** [comment]
This is very related to (and one possible solution for) https://github.com/pulumi/pulumi/issues/2059. 

The specific suggestion here has a challenge that it could only work for synchronous resource creations within the callback - which might be confusing. But that may still be acceptable, especially if combined with a way to globally force the default (which this callback could then build on top of to set and then unset the global default on entry/exit).

**Comment by tmehlinger** [comment]
Not sure if I follow you. If it's just a matter of the callback passed to the hypothetical method being [a-]synchronous, it could just as easily (in my mind, anyway 😄) look like this:
```typescript
const outputs = pulumi.withOptions({ provider: new SomeProvider() }, async () => {
    // make things!
});
```
That said, it's entirely possible (likely) I'm missing something here.

Thanks for the link to #2059. I followed that to https://github.com/pulumi/pulumi-kubernetes/issues/1517, which is actually where we're encountering this problem most often. My team would also like to have a nice dependency between k8s clusters and k8s cluster resources without having to jump through extra hoops. In the meantime, we can get around it by using the automation API to read the kubeconfig as a secret output from a cluster stack and populate `KUBECONFIG` in the environment before we kick off an app stack that depends on it.

Thanks for giving this some thought, and let me know how I can be of further help if you look more closely at this.

**Comment by jakzo** [comment]
I built something very similar to this before I saw this issue. It uses node's [AsyncLocalStorage](https://nodejs.org/api/async_context.html) which is stable as of node 16.

Usage:

```ts
export const resources = withStackTransformation(
  ({ props, opts }) => ({
    props,
    opts: { protect: true, ...opts },
  }),
  async () => {
    return {
      // Return any resources you want protected by default here...
    };
  }
);
```

Implementation:

```ts
import { AsyncLocalStorage } from "async_hooks";

import * as pulumi from "@pulumi/pulumi";

let _stackTrCtx:
  | AsyncLocalStorage<{
      transformations: pulumi.ResourceTransformation[];
    }>
  | undefined;
const getStackTransformContext = () => {
  const getStore = () => _stackTrCtx?.getStore() ?? { transformations: [] };
  if (_stackTrCtx) return { ctx: _stackTrCtx, store: getStore() };
  _stackTrCtx = new AsyncLocalStorage();
  pulumi.runtime.registerStackTransformation((args) => {
    for (const tr of getStore().transformations) {
      const result = tr(args);
      if (result) {
        args.props = result.props;
        args.opts = result.opts;
      }
    }
    return { props: args.props, opts: args.opts };
  });
  return { ctx: _stackTrCtx, store: getStore() };
};

export const withStackTransformation = <T>(
  transformation: pulumi.ResourceTransformation,
  callback: () => T
): T => {
  const { ctx, store } = getStackTransformContext();
  return ctx.run(
    { ...store, transformations: [...store.transformations, transformation] },
    callback
  );
};
```
